packages/amd-loader/amd-loader.js

```diff
@@ -0,0 +1,196 @@
+  this.resolved = false;
+  // Callbacks from dependents waiting to hear when this module has resolved.
+  this.notify = [];
+}
+
+var resolverAnchor = document.createElement('a');
+
+Module.prototype.init = function(deps, factory) {
+  var factoryArgs = [];
+  var numUnresolvedDeps = 0;
+
+  function onDepResolved() {
+    numUnresolvedDeps--;
+    maybeResolve();
+  }
+
+  var thisResolve = this.resolve.bind(this);
+  function maybeResolve() {
+    if (numUnresolvedDeps === 0) {
+      factory.apply(null, factoryArgs);
+      thisResolve();
+    }
+  }
+
+  for (var i = 0; i < (deps || []).length; i++) {
+    var depSpec = deps[i];
+
+    if (depSpec === 'exports') {
+      factoryArgs[i] = this.exports;
+
+    } else if (depSpec === 'require') {
+      factoryArgs[i] = this.makeDynamicImporter();
+
+    } else if (depSpec === 'meta') {
+      // TODO(aomarks) This is an idea for replacing the import.meta transform.
+      factoryArgs[i] = {url: this.url};
+
+    } else {
+      var depModule = registry.get(this.resolveUrl(depSpec));
+      factoryArgs[i] = depModule.exports;
+
+      if (depModule.resolved === false) {
+        numUnresolvedDeps++;
+        depModule.notify.push(onDepResolved);
+        depModule.loadOnce();
+      }
+    }
+  }
+
+  maybeResolve();
+};
+
+Module.prototype.resolve = function() {
+  this.resolved = true;
+  for (var i = 0; i < this.notify.length; i++) {
+    this.notify[i]();
+  }
+};
+
+Module.resolverAnchor = document.createElement('a');
+
+Module.prototype.resolveUrl = function(path) {
+  if (path.indexOf('://') !== -1) {
+    // Already a fully qualified URL.
+    // TODO(aomarks) Is this a good enough check?
+    return path;
+  }
+  // Just let the browser do path resolution/normalization.
+  Module.resolverAnchor.href = this.basePath + path;
+  return Module.resolverAnchor.href;
+};
+
+Module.prototype.makeDynamicImporter = function() {
+  var resolveUrl = this.resolveUrl.bind(this);
+  return function(spec, callback) {
+    var module = registry.get(resolveUrl(spec));
+    if (module.resolved === true) {
+      callback(module.exports);
+    } else {
+      module.notify.push(function() {
+        callback(module.exports);
+      });
+      module.loadOnce();
+    }
+  };
+};
+
+Module.prototype.loadOnce = function() {
+  if (this.loadStartedOrDone === true) {
+    return;
+  }
+  this.loadStartedOrDone = true;
+
+  var script = document.createElement('script');
+  script.src = this.url;
+
+  var module = this;
+  script.onload = function() {
+    if (registry.pendingDefine !== undefined) {
+      registry.pendingDefine(module);
+    } else {
+      // The script did not make a call to define(), otherwise the global
+      // callback would have been set. That's fine, we can resolve it
+      // immediately, because it doesn't have any dependencies.
+      module.resolve();
+    }
+  };
+
+  script.onerror = function() {
+    throw new Error('error loading module', url);
```

Comment:
does this need to propagate to dependents?

Reply:
Yes, and this now happens.